### PR TITLE
 feature request: allow functions as arguments in constant literals #1632 

### PR DESCRIPTION
**Pull request by e-hat**
This is an implementation for feature request #1632. The changes can be summarized as follows:
- Add variant Constant::Name (in compiler-core/src/ast/constant.rs)
- Parse identifiers into Constant::Name (in compiler-core/src/parse.rs)
- Parse function types for module constants (also in compiler-core/src/parse.rs)
- Typecheck/infer types for Constant::Name (in compiler-core/src/type_/expression.rs)
- Add metadata encoding/decoding for Constant::Var. This involves changing `schema.capnp` and `metadata/module_{encoder/decoder}.rs`
- Update how module constant statements are codegen'd in JS to properly output Constant::Name definition statements   (in `compiler-core/src/javascript/expression.rs`)
- Update how module constant usages are codegen'd in Elixir to properly output Constant::Name usages (in compiler-core/src/erlang.rs)
- Add tests for formatting, metadata encoding/decoding, and type checking. 

Should a language integration test be added for this as well? If so, how can I go about doing that?

### CODE REVIEW

**Review by lpil — COMMENTED** [review]
Great stuff! I've left a few questions inline.

Could we also get tests for Erlang generation, JavaScript generation, and that new error message? 🙏 

Thank you

File: compiler-core/src/type_/environment.rs
Comment (e-hat) [reply compiler-core/src/type_/environment.rs:310]: Perhaps. I added this because the following program was failing to compile:
```gleam
fn int_identity(i: Int) -> Int { i }
const f = int_identity
pub fn main() -> Int {
    f(42)
    ^ error here
}
```
The error originated inside `get_value_constructor`, where apparently `f` was not in `self.local_values`. Going up the stack, `get_value_constructor` was called by `get_field_map` which was called by `do_infer_call_with_known_fn`, called by `do_infer_call`. Inside `do_infer_call`, `f` is successfully inferred because `infer_value_constructor` searches the environment using both `get_variable` *and* `get_module_const`. I therefore assumed an additional check needed to be added to `get_value_constructor` to look in `module_values`. Is this a bug?

File: compiler-core/src/error.rs
Comment (e-hat) [reply compiler-core/src/error.rs:1930]: Hm...how about `Expected a function or a module constant`?

I thought I mentioned this somewhere but I may have forgotten -- I'm 99% sure it's impossible for the user to ever cause this error, so maybe it could be removed and replaced with an `unreachable` statement (unless that's bad style, I'm new to rust). The reasons for why it is impossible to cause are a little convoluted, so maybe we should just keep it in case something changes somewhere.

File: compiler-core/src/type_/environment.rs
Comment (e-hat) [reply compiler-core/src/type_/environment.rs:310]: While I would love to help, it might be better for you to do this refactoring.

Comment (lpil) [reply compiler-core/src/type_/environment.rs:310]: Righto, I'll try and get to this soon.

Comment (lpil) [reply compiler-core/src/type_/environment.rs:310]: @e-hat all done! https://github.com/gleam-lang/gleam/commit/49ef2769629db3316dcd225aef2fa7cd4d6b324e

**Comment by e-hat** [comment]
I added some codegen tests for this. I wanted to ask for some guidance before I write the test for the error message -- it's currently impossible to write a gleam program that causes this error message. 

Should I try to construct this error artificially in the test or should we be handling the situation that causes this error differently?

File: compiler-core/src/error.rs
Comment (lpil) [reply compiler-core/src/error.rs]: Ah nice! Let's use `unreachable` with a description of why we think that is the case.

**Comment by lpil** [comment]
@e-hat Really sorry for the delay with this one, I only just saw your comment above. I have replied now

**Comment by e-hat** [comment]
Ok, everything should be good to go now. Let me know if we need anything else.

File: compiler-core/src/javascript/tests/functions.rs
Comment (lpil) [inline compiler-core/src/javascript/tests/functions.rs:371]: What happens if the const is higher up in the file than the function it references? Does it still generate valid to JavaScript?

**Review by lpil — COMMENTED** [review]
Looking really good! There is one small bug and I have left another question but after that I think this is ready to go. Thanks

File: compiler-core/src/javascript/tests/functions.rs
Comment (e-hat) [reply compiler-core/src/javascript/tests/functions.rs:371]: Good catch. It seems like when you assign a `Const::Var` to another module constant, it doesn't work.
```gleam
const x = y // this fails, saying that `y` is not in scope
const y = 5
```
However, when you assign a `Const::Var` to module function, it *does* work.
```gleam
const f_alias = f // this works
fn f(i: Int) -> Int { i }
```
This discrepancy feels like it could be a little confusing. I'm not sure how that part of the code works, but I assume that when a module is type checked it first sweeps through the function signatures so that you can use them out of order. Would it be difficult to add the module constants to this initial scope as well? I'd love to hear what you think about this.

I also think it could be error-prone to simply rely on the underlying behavior of javascript/erlang/whatever to be able to handle out of order usages/declarations of these global values. This is what this PR does currently -- it just declares the constants in the emitted code in the same order as they appear in the source.

Comment (lpil) [reply compiler-core/src/javascript/tests/functions.rs:371]: There is an issue for the ordering of constants limitation, so I don't think we need to worry about that here. We do need to make it so that this feature will always output valid JavaScript though, so we will need to do more than just inmates the statements in the order that we get them.

File: compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__variables__module_const_vars.snap
Comment (e-hat) [reply compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__variables__module_const_vars.snap:20]: Which part is a syntax error? I'm not super familiar with erlang's syntax. This erlang code compiles and runs on my computer.

File: compiler-core/src/javascript/tests/functions.rs
Comment (e-hat) [reply compiler-core/src/javascript/tests/functions.rs:371]: Ok, so if the ordering of the constants isn't an issue, then I think we always output valid JS. The case you originally asked about, referring to a function in a constant definition before that function has been defined, does output valid JS as long as the JS codegen outputs functions using the `function` keyword due to hoisting. Is there an edge case that I'm missing?

File: compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__variables__module_const_vars.snap
Comment (htstudios) [reply compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__variables__module_const_vars.snap:20]: `int_identity(42)`

Comment (inoas) [reply compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__variables__module_const_vars.snap:20]: `fun int_identity/1(42).` => `int_identity(42).`

Comment (e-hat) [reply compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__variables__module_const_vars.snap]: Ok, I think [754ab16](https://github.com/gleam-lang/gleam/pull/1664/commits/754ab168d5ceebc47d4b5e1b5a29a7e40d550551) should fix this. Had to add a special match for the case when we are emitting a function call and the function is a `Constant::Var` that contains a `ValueConstructorVariant::ModuleFn` so that this could be handled the same as a normal function call.

**Comment by TanklesXL** [comment]
What's going on with the CI for the PR? All I see is it's waiting for status to be reported, does it need approval to run again?

**Comment by lpil** [comment]
@TanklesXL it has a merge conflict so CI can't run. I will resolve that

File: compiler-core/src/javascript/tests/functions.rs
Comment (lpil) [reply compiler-core/src/javascript/tests/functions.rs:371]: Brilliant, thank you. No edge case as far as I'm aware, it was a genuinely question as I was unsure.

**Review by lpil — APPROVED** [review]
Brilliant work, thanks once again. And thank you for being so patient with me as I catch up on Gleam PRs.

**Comment by TanklesXL** [comment]
Thanks a lot everyone, so excited for this! ❤️ 🚀

**Pull request merged**
